text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+dictionary and schema information.  This remains an area for study.
+
+## Appendix A - Protocol Buffer Definitions
+
+Protobuf specification for an Arrow-based OpenTelemetry event.
+
+```protobuf
+syntax = "proto3";
+
+package opentelemetry.proto.collector.arrow.v1;
+
+option java_multiple_files = true;
+option java_package = "io.opentelemetry.proto.collector.arrow.v1";
+option java_outer_classname = "ArrowServiceProto";
+
+// Note the following is temporary
+option go_package = "github.com/f5/otel-arrow-adapter/api/collector/arrow/v1";
+
+service ArrowStreamService {
+  // The ArrowStream endpoint is a bi-directional stream used to send batch of `BatchArrowRecords` from the exporter
+  // to the collector. The collector returns `BatchStatus` messages to acknowledge the `BatchArrowRecords` messages received.
+  rpc ArrowStream(stream BatchArrowRecords) returns (stream BatchStatus) {}
+}
+
+// A message sent by an exporter to a collector containing a batch of Arrow records.
+message BatchArrowRecords {
+  // [mandatory] Batch ID. Must be unique in the context of the stream.
+  string batch_id = 1;
+
+  // [mandatory] A collection of payloads containing the data of the batch.
+  repeated OtlpArrowPayload otlp_arrow_payloads = 2;
+}
+
+// Enumeration of all the OTLP Arrow payload types currently supported by the OTLP Arrow protocol.
+enum OtlpArrowPayloadType {
+  // A payload representing a collection of metrics.
+  METRICS = 0;
+  // A payload representing a collection of logs.
+  LOGS = 1;
+  // A payload representing a collection of traces.
+  SPANS = 2;
+}
+
+// Represents a batch of OTLP Arrow entities.
+message OtlpArrowPayload {
+  // [mandatory] A unique id assigned to a sub-stream of the batch sharing the same schema, and dictionaries.
+  string sub_stream_id = 1;
+
+  // [mandatory] Type of the OTLP Arrow payload.
+  OtlpArrowPayloadType type = 2;
+
+  // [mandatory] Serialized Arrow Record Batch
+  // For a description of the Arrow IPC format see: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+  bytes record = 4;
+
+  // [mandatory]
+  CompressionMethod compression = 5;
+}
+
+// The compression method used to compress the different bytes buffer.
+enum CompressionMethod {
+  NO_COMPRESSION = 0;
+  ZSTD = 1;
+}
+
+// Arrow IPC message
+// see: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+message EncodedData {
+  // Serialized Arrow encoded IPC message
+  bytes ipc_message = 1;
+
+  // Serialized Arrow buffer
+  bytes arrow_data = 2;
+}
+
+// A message sent by a Collector to the exporter that opened the Jodata stream.
+message BatchStatus {
+  repeated StatusMessage statuses = 1;
+}
+
+message StatusMessage {
+  string batch_id = 1;
+  StatusCode status_code = 2;
+  ErrorCode error_code = 3;
+  string error_message = 4;
+  RetryInfo retry_info = 5;
```

Comment:
I think we should keep this field, but possibly we could refer directly to a gRPC details type that encodes the information `exporterhelper` uses to determine retries? (cc @tigrannajaryan)